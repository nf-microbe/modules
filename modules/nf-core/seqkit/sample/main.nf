process SEQKIT_SAMPLE {
    tag "${meta.id}"
    label 'process_low'
    // File IO can be a bottleneck. See: https://bioinf.shenwei.me/seqkit/usage/#parallelization-of-cpu-intensive-jobs

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/seqkit:2.8.1--h9ee0642_0':
        'biocontainers/seqkit:2.8.1--h9ee0642_0' }"

    input:
    tuple val(meta), path(fastx)

    output:
    tuple val(meta), path("${prefix}*") , emit: fastx
    path "versions.yml"                 , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args        = task.ext.args ?: ''
    def args2       = task.ext.args2 ?: ''
    prefix          = task.ext.prefix ?: "${meta.id}"
    def extension   = "fastq"
    if ("$fastx" ==~ /.+\.fasta|.+\.fasta.gz|.+\.fa|.+\.fa.gz|.+\.fas|.+\.fas.gz|.+\.fna|.+\.fna.gz|.+\.fsa|.+\.fsa.gz/ ) {
        extension   = "fasta"
    }
    extension       = fastx.toString().endsWith('.gz') ? "${extension}.gz" : extension
    def call_gzip   = extension.endsWith('.gz') ? "| gzip -c $args2" : ''
    if("${prefix}.${extension}" == "$fastx") error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
    if (meta.single_end || meta.assembler != null) {
        """
        seqkit \\
            sample \\
            --threads ${task.cpus} \\
            ${args} \\
            ${fastx} \\
            ${call_gzip} \\
            > ${prefix}.${extension}

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            seqkit: \$(seqkit version | cut -d' ' -f2)
        END_VERSIONS
        """
    } else {
        """
        seqkit \\
            sample \\
            --threads ${task.cpus} \\
            ${args} \\
            ${fastx[0]} \\
            --rand-seed 1 \\
            ${call_gzip} \\
            > ${prefix}_1.${extension}

        seqkit \\
            sample \\
            --threads ${task.cpus} \\
            ${args} \\
            ${fastx[1]} \\
            --rand-seed 1 \\
            ${call_gzip} \\
            > ${prefix}_2.${extension}

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            seqkit: \$(seqkit version | cut -d' ' -f2)
        END_VERSIONS
        """
    }

    stub:
    prefix          = task.ext.prefix ?: "${meta.id}"
    def extension   = "fastq"
    if ("$fastx" ==~ /.+\.fasta|.+\.fasta.gz|.+\.fa|.+\.fa.gz|.+\.fas|.+\.fas.gz|.+\.fna|.+\.fna.gz|.+\.fsa|.+\.fsa.gz/ ) {
        extension   = "fasta"
    }
    extension = fastx.toString().endsWith('.gz') ? "${extension}.gz" : extension
    if("${prefix}.${extension}" == "$fastx") error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
    def is_compressed = extension.endsWith(".gz") ? true : false
    if (meta.single_end || meta.assembler != null) {
        """
        if [ "$is_compressed" == "true" ]; then
            echo "" | gzip > ${prefix}.${extension}
        else
            touch ${prefix}.${extension}
        fi

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            seqkit: \$(seqkit version | cut -d' ' -f2)
        END_VERSIONS
        """
    } else {
        """
        if [ "$is_compressed" == "true" ]; then
            echo "" | gzip > ${prefix}_1.${extension}
            echo "" | gzip > ${prefix}_2.${extension}
        else
            touch ${prefix}_1.${extension}
            touch ${prefix}_2.${extension}
        fi

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            seqkit: \$(seqkit version | cut -d' ' -f2)
        END_VERSIONS
        """
    }
}
