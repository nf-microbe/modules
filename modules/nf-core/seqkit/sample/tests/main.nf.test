nextflow_process {

    name "Test Process SEQKIT_SAMPLE"
    script "../main.nf"
    process "SEQKIT_SAMPLE"
    config './nextflow.config'

    tag "modules"
    tag "modules_nfcore"
    tag "seqkit"
    tag "seqkit/sample"

    test("sarscov2-genome_fasta") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test', assembler:'megahit' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
            )
        }
    }

    test("sarscov2-genome_fasta_gz") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test', assembler:'megahit'  ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta.gz', checkIfExists: true)
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
            )
        }
    }

    test("sarscov2-fastq pe") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true)
                    ]
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
            )
        }
    }

    test("sarscov2-test_1_fastq_gz se") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:true ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true)
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
            )
        }
    }

    test("sarscov2-genome_fasta-stub") {
        options '-stub'
        when {
            process {
                """
                input[0] = [
                    [ id:'test', assembler:'megahit'  ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
            )
        }
    }

    test("sarscov2-fastq pe - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true)
                    ]
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
            )
        }
    }
}
