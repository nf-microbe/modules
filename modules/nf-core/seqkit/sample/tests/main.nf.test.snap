{
    "sarscov2-genome_fasta-stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.fasta:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,c1281e3f2fe960a95147ff8efe52986a"
                ],
                "fastx": [
                    [
                        {
                            "id": "test"
                        },
                        "test.fasta:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c1281e3f2fe960a95147ff8efe52986a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-11T11:30:40.03141103"
    },
    "sarscov2-test_1_fastq_gz": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.fastq.gz:md5,3dd77b8b832bb977908008619eeab7be"
                    ]
                ],
                "1": [
                    "versions.yml:md5,c1281e3f2fe960a95147ff8efe52986a"
                ],
                "fastx": [
                    [
                        {
                            "id": "test"
                        },
                        "test.fastq.gz:md5,3dd77b8b832bb977908008619eeab7be"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c1281e3f2fe960a95147ff8efe52986a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-11T11:30:19.62092629"
    },
    "sarscov2-genome_fasta": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.fasta:md5,483f4a5dfe60171c86ee9b7e6dff908b"
                    ]
                ],
                "1": [
                    "versions.yml:md5,c1281e3f2fe960a95147ff8efe52986a"
                ],
                "fastx": [
                    [
                        {
                            "id": "test"
                        },
                        "test.fasta:md5,483f4a5dfe60171c86ee9b7e6dff908b"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c1281e3f2fe960a95147ff8efe52986a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-11T11:29:49.532001208"
    },
    "sarscov2-genome_fasta_gz": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.fasta.gz:md5,483f4a5dfe60171c86ee9b7e6dff908b"
                    ]
                ],
                "1": [
                    "versions.yml:md5,c1281e3f2fe960a95147ff8efe52986a"
                ],
                "fastx": [
                    [
                        {
                            "id": "test"
                        },
                        "test.fasta.gz:md5,483f4a5dfe60171c86ee9b7e6dff908b"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c1281e3f2fe960a95147ff8efe52986a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-11T11:30:04.665415595"
    }
}